filters/NeighborClassifierFilter.cpp

```diff
@@ -0,0 +1,199 @@
+* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+* COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+* OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+* OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+* OF SUCH DAMAGE.
+****************************************************************************/
+
+#include "NeighborClassifierFilter.hpp"
+
+#include <pdal/PipelineManager.hpp>
+#include <pdal/StageFactory.hpp>
+#include <pdal/util/ProgramArgs.hpp>
+
+#include "private/DimRange.hpp"
+
+#include <iostream>
+#include <utility>
+namespace pdal
+{
+
+static PluginInfo const s_info = PluginInfo(
+    "filters.neighborclassifier",
+    "Re-assign some point attributes based KNN voting",
+    "http://pdal.io/stages/filters.neighborclassifier.html" );
+
+CREATE_STATIC_PLUGIN(1, 0, KNNAssignFilter, Filter, s_info)
+
+KNNAssignFilter::KNNAssignFilter() : m_dim(Dimension::Id::Classification)
+{}
+
+
+KNNAssignFilter::~KNNAssignFilter()
+{}
+
+
+void KNNAssignFilter::addArgs(ProgramArgs& args)
+{
+    args.add("domain", "Selects which points will be subject to KNN-based assignmenassignment",
+        m_domainSpec);
+    args.add("k", "Number of nearest neighbors to consult",
+        m_k).setPositional();
+    //args.add("dimension", "Dimension on to be updated", m_dimName).setPositional();
+    Arg& candidate = args.add("candidate", "candidate file name",
+        m_candidateFile);
+}
+
+void KNNAssignFilter::initialize()
+{
+    for (auto const& r : m_domainSpec)
+    {
+        try
+        {
+            DimRange range;
+            range.parse(r);
+            m_domain.push_back(range);
+        }
+        catch (const DimRange::error& err)
+        {
+            throwError("Invalid 'domain' option: '" + r + "': " + err.what());
+        }
+    }
+    if (m_k < 1)
+        throwError("Invalid 'k' option: " + std::to_string(m_k) +  ", must be > 0");
+}
+void KNNAssignFilter::prepared(PointTableRef table)
+{
+    PointLayoutPtr layout(table.layout());
+
+    for (auto& r : m_domain)
+    {
+        r.m_id = layout->findDim(r.m_name);
+        if (r.m_id == Dimension::Id::Unknown)
+            throwError("Invalid dimension name in 'domain' option: '" +
+                r.m_name + "'.");
+    }
+    std::sort(m_domain.begin(), m_domain.end());
+    //m_dim = layout->findDim(m_dimName);
+
+    //if (m_dim == Dimension::Id::Unknown)
+    //    throwError("Dimension '" + m_dimName + "' not found.");
+}
+
+void KNNAssignFilter::doOneNoDomain(PointRef &point, PointRef &temp, KD3Index &kdi)
+{
+    std::vector<PointId> iSrc = kdi.neighbors(point, m_k);
```

Comment:
In the case that the query point and the KD3Index are from the same point cloud, `neighbors` will return the query point. If that is not the intended behavior, you could detect the situation and increment the number of neighbors accordingly.

Reply:
Let's see ... I had two use cases in mind:
1.  A single point cloud where we need to smooth the classifications, so K > 1 (think a few stray mis-classifications)
2.  Two point clouds, the candidate is completely classified and the source (where the query point comes from) is a superset of those points.  We need to extrapolate the classifications, from the smaller candidate to the larger source.  Here K == 1.